javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java

```diff
@@ -587,36 +605,143 @@ private boolean verifyYarnInstallation() {
+  protected void restoreOriginalFiles(Map<Path, Path> originalFiles) {
+    originalFiles.forEach(
+        (file, original) -> {
+          try {
+            Files.move(original, file, StandardCopyOption.REPLACE_EXISTING);
+          } catch (IOException e) {
+            throw new ResourceError("Could not restore original file: " + file, e);
+          }
+        });
+  }
+
+  private static final Pattern validPackageName = Pattern.compile("(@[\\w.-]+/)?\\w[\\w.-]*");
+
+  protected Map<Path, Path> installDependencies(Set<Path> filesToExtract) {
     if (!verifyYarnInstallation()) {
-      return;
+      return Collections.emptyMap();
     }
+
+    Path rootNodeModules = Paths.get("node_modules");
+
+    // Read all package.json files, and install symlinks to them.
+    Map<Path, JsonObject> packageJsonFiles = new LinkedHashMap<>();
+    Set<String> packagesInRepo = new LinkedHashSet<>();
     for (Path file : filesToExtract) {
       if (file.getFileName().toString().equals("package.json")) {
-        System.out.println("Installing dependencies from " + file);
-        ProcessBuilder pb =
-            new ProcessBuilder(
-                Arrays.asList(
-                    "yarn",
-                    "install",
-                    "--non-interactive",
-                    "--ignore-scripts",
-                    "--ignore-platform",
-                    "--ignore-engines",
-                    "--ignore-optional",
-                    "--no-default-rc",
-                    "--no-bin-links",
-                    "--pure-lockfile"));
-        pb.directory(file.getParent().toFile());
-        pb.redirectOutput(Redirect.INHERIT);
-        pb.redirectError(Redirect.INHERIT);
         try {
-          pb.start().waitFor(this.installDependenciesTimeout, TimeUnit.MILLISECONDS);
-        } catch (IOException | InterruptedException ex) {
-          throw new ResourceError("Could not install dependencies from " + file, ex);
+          String text = new WholeIO().read(file);
+          JsonElement json = new JsonParser().parse(text);
+          if (!(json instanceof JsonObject)) continue;
+          JsonObject jsonObject = (JsonObject) json;
+          file = file.toAbsolutePath();
+          packageJsonFiles.put(file, jsonObject);
+
+          JsonElement nameElm = jsonObject.get("name");
+          if (nameElm instanceof JsonPrimitive && ((JsonPrimitive) nameElm).isString()) {
+            String name = nameElm.getAsString();
+            packagesInRepo.add(name);
+
+            if (validPackageName.matcher(name).matches()) {
+              // Create a symlink to the package: <checkout>/node_modules/foo -> /path/to/foo
+              try {
+                Path symlinkPath = rootNodeModules.resolve(name);
+                if (!Files.exists(symlinkPath)) {
+                  Files.createDirectories(symlinkPath.getParent());
+                  Files.createSymbolicLink(symlinkPath, file.getParent());
```

Comment:
Is this going to work on Windows? [StackOverflow](https://stackoverflow.com/questions/23217460/how-to-create-soft-symbolic-link-using-java-nio-files) suggests that one needs special permissions to create symbolic links on Windows. Not an issue for LGTM.com, of course, but still relevant for LGTM-E.

Reply:
That's a good question. I'm having some difficulty getting Java to run on my Windows VM today but I'll try to test it this week.

FWIW yarn workspaces use symlinks and yarn manages to create them just fine on Windows, so maybe can use that as a fallback.

Reply:
Turns out yarn workspaces use "junctions", a type of symlink on NTFS which doesn't require admin. The Java API doesn't have a method for creating them.

Before going any further with the symlink approach, I'll experiment a bit with providing in a virtual file system in the compiler host we pass to the TypeScript compiler.